thorlcr/activities/indexread/thindexreadslave.cpp

```diff
@@ -168,15 +168,65 @@ class CIndexReadSlaveBase : public CSlaveActivity
                 IOutputMetaData *actualFormat = translator ? &translator->queryActualFormat() : expectedFormat;
                 bool tryRemoteStream = actualFormat->queryTypeInfo()->canInterpret() && actualFormat->queryTypeInfo()->canSerialize() &&
                                        projectedFormat->queryTypeInfo()->canInterpret() && projectedFormat->queryTypeInfo()->canSerialize();
-                bool usesBlobs = 0 != (helper->getFlags() & TIRusesblob);
+
+                offset_t avgPartSize = 0;
+                offset_t keyCacheSize = queryJob().getKeyNodeCacheSize() + queryJob().getKeyLeafCacheSize();
+                auto keyCacheRatioHeuristic = [keyCacheSize](offset_t size, unsigned numParts, unsigned compressionMultiple, unsigned cacheSizeMultiple)
+                {
+                    /*
+                     * Rough heuristic.
+                     *
+                     * If ([average compressed part size] * [# parts handling] * [compressionMultiple] < [cacheSizeMultiple]
+                     *
+                     * If the total size of compressed index data * [compressionMultiple - for rough estimate of expanded size]
+                     * is less than the slaves key cache size * [cacheSizeMultiple],
+                     * then disabled remote streaming.
+                     *
+                     * cacheSizeMultiple because still want to turn off if hitting it e.g. only hitting cache a 1/4 of the time.
+                     *
+                     */
+
+                    return ((size * numParts) * compressionMultiple) < (keyCacheSize * cacheSizeMultiple);
+                };
+
+                if (0 != (helper->getFlags() & TIRusesblob))
+                    tryRemoteStream = false;
+                else if (localMerge)
+                    tryRemoteStream = false;
+                else if (inChildQuery() && !getOptBool(THOROPT_STREAM_REMOTE_SMALL_INDEX)) // option turns of heuristic
+                {
+                    /*
+                     * If in a CQ, it is counterproductive to use an index read stream per CQ execution if the index
+                     * involved is relatively small.
+                     * Because, if the index is small and direct reading (and caching) key node pages, it is likely
+                     * that repeated executions will not read any (or few) new key pages (i.e. cache hit).
+                     *
+                     * Example: small 1-way key being remotely read by the whole cluster.
+                     * If it is small it will fit (or mostly fit) in node key cache, and thus mostly read from memory vs over the network etc.
+                     *
+                     */
+                    offset_t logicalFileSize = part.queryOwner().queryProperties().getPropInt64("@size"); // NB: size is compressed size
```

Comment:
Shouldn't this be calculated once before reading the indexes, rather than on each file part?

Reply:
yes agree, I'll move the calculation so done once.